apps/rush-lib/src/logic/PackageChangeAnalyzer.ts

```diff
@@ -92,7 +102,13 @@ export class PackageChangeAnalyzer {
         | RushConfigurationProject
         | undefined = this._rushConfiguration.findProjectForPosixRelativePath(filePath);
       if (owningProject) {
-        projectHashDeps.get(owningProject.packageName)!.set(filePath, fileHash);
+        const relativePath: string = filePath
+          .replace(owningProject.projectRelativeFolder, '')
+          .replace(/^\//, '');
```

Comment:
Could you add a code comment explaining what transformation is happening here?  

It might be helpful to include a realistic example of the input/output, something like:

```ts
        // Example:
        // filePath: category\project\src\thing.ts
        // projectRelativeFolder: category\project
        // relativePath: src\thing.ts
        const relativePath: string = filePath
          .replace(owningProject.projectRelativeFolder, '')
          .replace(/^\//, '');
```

It seems obvious to me now, but I had to run this code in the debugger to figure out what it is doing. :-)

Reply:
The correctness of `replace()` hinges on the assumption that `filePath` will always contain `projectRelativeFolder`, but it's difficult to prove that, since `filePath` and `projectRelativeFolder` come from fairly different subsystems.

Here's a couple ideas to protect against that:
1. Use `path.relative()` to operate on path objects rather than text strings; OR
2. First test whether `filePath` starts with `projectRelativeFolder` and throw an `InternalError` if that not. The assumption is probably safe, but if not it would be nice to hear about it so we can fix it.

(BTW I tried some experiments to see if I could break this logic, like altering the case of a project folder on Windows. But the `LookupByPath` API malfunctions first preventing this code from running. Maybe I'm being overly defensive -- it's just if `replace()` did fail to match, the behavior might be pretty difficult to understand. 🙂)

@dmichon-msft

Reply:
The base logic here is that, since `LookupByPath` cannot return a matching project for a path that does not literally begin with `projectRelativeFolder`, all paths must necessarily begin with said string prefix. Personally I'd just blindly slice the expected length off of the start of the string, because that is how I prefer to do path manipulation (as quickly as possible, with a minimum of fuss).
Actually the `.slice()` call is safer here, because, contrary to what my memory had said, I only normalize the `projectRelativeFolder` to `/` when handing it off to `LookupByPath`, so you can technically (incorrectly) use `\\` inside of your paths in rush.json and weird things will happen.

Reply:
I've split the difference by adding a comment _and_ converting it to use @dmichon-msft 's suggested logic.

Reply:
👍 I am content that sufficient paranoia has been applied here heheh